File: leetcode/c++/easy/_4_median_of_two_sorted_arrays.cpp
// Leetcode Problem 4

#include <gtest/gtest.h>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>&nums2) {
        vector<int> result;
        result.reserve(nums1.size() + nums2.size());

        merge(nums1.begin(), nums1.end(),
              nums2.begin(), nums2.end(),
              back_inserter(result));

        if (result.size() % 2 == 0) {
            return (result[result.size() / 2] + result[result.size() / 2 - 1]) / 2.0;
        } else {
            return (double)(result[result.size() / 2]);
        }
    }
};


class TestFindMedian : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestFindMedian, TestCase1) {
    vector<int> nums1 = {1, 3};
    vector<int> nums2 = {2};

    EXPECT_EQ(solution.findMedianSortedArrays(nums1, nums2), 2.0);
}

TEST_F(TestFindMedian, TestCase2) {
    vector<int> nums1 = {1, 2};
    vector<int> nums2 = {3, 4};

    EXPECT_EQ(solution.findMedianSortedArrays(nums1, nums2), 2.5);
}

File: leetcode/c#/easy/_4_median_of_two_sorted_arrays.cs
// Leetcode Problem 4

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace easy;

public class Solution_FindMedianSortedArrays {
    public double FindMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] result = nums1.Concat(nums2).ToArray();
        Array.Sort(result);

        if (result.Length % 2 == 0) {
            return (result[result.Length / 2] + result[result.Length / 2 - 1]) / 2.0;
        } else {
            return result[result.Length / 2];
        }
    }

    [TestClass]
    public class TestFindMedianSortedArrays {
        private Solution_FindMedianSortedArrays solution;

        [TestInitialize]
        public void Setup() {
            solution = new Solution_FindMedianSortedArrays();
        }

        [TestMethod]
        public void TestFindMedianSortedArrays_SimpleTestCase1() {
            int[] nums1 = [1, 3];
            int[] nums2 = [2];
            double expected = 2.0;

            Assert.AreEqual(expected, solution.FindMedianSortedArrays(nums1, nums2));
        }

        [TestMethod]
        public void TestFindMedianSortedArrays_SimpleTestCase2() {
            int[] nums1 = [1, 2];
            int[] nums2 = [3, 4];
            double expected = 2.5;

            Assert.AreEqual(expected, solution.FindMedianSortedArrays(nums1, nums2));
        }
    }
}